sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,463 @@
+additional kernels without involving the host program. In OpenCL 2.0 and beyond,
+kernel invocations can be initiated without host program interaction through
+the `enqueue_kernel` API. C++11 also introduced `std::async()` for task-level
+parallelism. However, neither of these approaches is a good fit for spatial
+compute platforms, which require a clear mapping of parallel constructs to
+spatial instances in the hardware.
+
+Rather than significantly changing the semantics of existing solutions, this
+extension introduces a sub-kernel task-level parallelism interface to SYCL for
+use in device code. It provides the class
+`sycl::ext::intel::experimental::task_sequence`, which is an
+abstraction of an asynchronous sequence of invocations of a callable that can be
+executed asynchronously from the caller.  A single object of this class can map
+naturally to a single hardware instance on spatial targets.
+
+NOTE: In this document, we use `task_sequence` to indicate the proposed
+`sycl::ext::intel::experimental::task_sequence`
+
+=== Examples
+
+Here's an example that shows how to define and call two task sequences.
+
+```c++
+using namespace sycl::ext::intel::experimental;
+
+// Return the sum of elements in 'v', from index 's' to 's+sz'
+int vectorSum(float *v, size_t s, size_t sz) {
+  ...
+}
+
+{
+  sycl::queue q{...};
+
+  constexpr kCount = ...;
+  std::vector<float> in(kCount);
+  float* out = ...;
+  ...
+
+  q.single_task([=] {
+    task_sequence<vectorSum> firstQuarter;
+    task_sequence<vectorSum> secondQuarter;
+    task_sequence<vectorSum> thirdQuarter;
+    task_sequence<vectorSum> fourthQuarter;
+    int quarterCount = kCount/4;
+    firstQuarter.async(in.data(), 0, quarterCount);
+    secondQuarter.async(in.data(), quarterCount, quarterCount);
+    thirdQuarter.async(in.data(), 2*quarterCount, quarterCount);
+    fourthQuarter.async(in.data(), 3*quarterCount, quarterCount);
+    *out = firstQuarter.get() + secondQuarter.get() +
+      thirdQuarter.get() + fourthQuarter.get();
+  }).wait();
+  ...
+}
+```
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_FPGA_TASK_SEQUENCE` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+== Definitions
+
+== `task_sequence` class
+
+[source,c++,linenums]
+----
+namespace sycl::ext::intel::experimental {
+
+template<auto& f, class propertiesT =
+   decltype(oneapi::experimental::properties{})>
+class task_sequence{};
```

Comment:
Let's make it an compile-time error to match this template instead of the partial one below.  That could be accomplished with a static_assert + a check class like this:

```
template<auto &f>
struct task_sequence_checker : std::false_type {};

template<typename ReturnT, typename... ArgsT, ReturnT (&f)(ArgsT...)>
struct task_sequence_checker<f> : std::true_type {};

template<auto& f, class propertiesT =
   decltype(oneapi::experimental::properties{})>
class task_sequence{
  static_assert(task_sequence_checker<f>::value);
};

Reply:
Added, thank you! :)

Reply:
This was suggested to be removed since it is an implementation detail and not part of the API.